velox/core/PlanNode.cpp

```diff
@@ -504,6 +504,18 @@ void addWindowFunction(
   if (frame.endType == WindowNode::BoundType::kUnboundedPreceding) {
     VELOX_USER_FAIL("Window frame end cannot be UNBOUNDED PRECEDING");
   }
+  if (frame.type == WindowNode::WindowType::kRange) {
+    if (frame.startType == WindowNode::BoundType::kPreceding ||
+        frame.startType == WindowNode::BoundType::kFollowing) {
+      VELOX_NYI(
+          "k preceding/following as frame start is allowed only in ROWS mode");
```

Comment:
Its sufficient to say that "k preceding/following is only supported in ROWS mode" and put all the cases in a single condition.

Reply:
On second thoughts, do you want to put this error here. It might be better to put it only in the Window operator since the PlanBuilder tests wouldn't need to be modified,